File: ThucHanh/Timerapp/app/src/main/java/com/example/timerapp/MainActivity.java
package com.example.timerapp;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {

    private TextView timeTextView;
    private Handler handler;
    private int secondsElapsed = 0;
    private boolean isRunning = true;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Ánh xạ TextView
        timeTextView = findViewById(R.id.timeTextView);

        // Khởi tạo Handler gắn với Main Looper (UI Thread)
        handler = new Handler(Looper.getMainLooper());

        // Tạo và chạy thread đếm giờ
        startTimerThread();
    }

    private void startTimerThread() {
        // Tạo Runnable để cập nhật UI
        Runnable updateUIRunnable = new Runnable() {
            @Override
            public void run() {
                // Cập nhật TextView trên UI Thread
                timeTextView.setText("Time elapsed: " + secondsElapsed + " seconds");

                // Lên lịch cho lần cập nhật tiếp theo sau 1 giây
                if (isRunning) {
                    handler.postDelayed(this, 1000);
                }
            }
        };

        // Tạo thread nền để đếm thời gian
        Thread timerThread = new Thread(new Runnable() {
            @Override
            public void run() {
                while (isRunning) {
                    try {
                        // Tăng thời gian mỗi giây
                        Thread.sleep(1000);
                        secondsElapsed++;

                        // Gửi Runnable tới Handler để cập nhật UI
                        handler.post(updateUIRunnable);
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        });

        // Bắt đầu thread
        timerThread.start();

        // Đăng ký lần cập nhật UI đầu tiên
        handler.post(updateUIRunnable);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        // Dừng thread khi activity bị hủy
        isRunning = false;
        handler.removeCallbacksAndMessages(null); // Xóa tất cả callback từ Handler
    }
}

File: ThucHanh/StopwatchApp/app/src/main/java/com/example/stopwatchapp/MainActivity.kt
package com.example.stopwatchapp

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity

class MainActivity : AppCompatActivity() {

    private lateinit var tvTimer: TextView
    private lateinit var btnStart: Button
    private lateinit var btnStop: Button

    private var timeElapsed = 0 // Biến đếm giây
    private var isRunning = false // Kiểm tra trạng thái bộ đếm

    private val handler = Handler(Looper.getMainLooper()) // Handler chạy trên UI thread
    private val runnable = object : Runnable {
        override fun run() {
            if (isRunning) {
                timeElapsed++
                tvTimer.text = timeElapsed.toString()
                handler.postDelayed(this, 1000) // Lặp lại sau 1 giây
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        tvTimer = findViewById(R.id.tvTimer)
        btnStart = findViewById(R.id.btnStart)
        btnStop = findViewById(R.id.btnStop)

        btnStart.setOnClickListener {
            if (!isRunning) {
                isRunning = true
                handler.post(runnable) // Bắt đầu đếm giờ
            }
        }

        btnStop.setOnClickListener {
            isRunning = false
            handler.removeCallbacks(runnable) // Dừng đếm giờ
        }
    }
}
